drivers/include/adc_ng.h

```diff
@@ -0,0 +1,242 @@
+
+#include "adc_ng_internal.h"
+#include "bitarithm.h"
+#include "periph_conf.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Pass this special value as parameter `ref` in @ref adc_ng_init
+ *          to select the highest supported reference voltage
+ */
+#define ADC_NG_MAX_REF                  (0U)
+
+/**
+ * @name    Helper functions query supported ADC resolutions
+ *
+ * @{
+ */
+
+/**
+ * @brief   Check if the given ADC supports the given resolution
+ *
+ * @param[in]       adc     ADC device to check
+ * @param[in]       res     Resolution to check
+ *
+ * @retval  1               Resolution is supported
+ * @retval  0               Resolution is not supported
+ */
+static inline int adc_ng_supports_res(uint8_t adc, uint8_t res)
+{
+    assert(adc < ADC_NG_NUMOF);
+    assert(res < 32);
+    return ((res < 32) && (adc_ng_drivers[adc]->res_supported & (1 << res)));
+}
+
+/**
+ * @brief   Get the highest supported resolution of an ADC
+ *
+ * @param[in]       adc     ADC to get the highest supported resolution of
+ */
+static inline uint8_t adc_ng_max_res(uint8_t adc)
+{
+    assert(adc < ADC_NG_NUMOF);
+    return (uint8_t)bitarithm_msb(adc_ng_drivers[adc]->res_supported);
+}
+
+/**
+ * @brief   Get the highest supported resolution of an ADC
+ *
+ * @param[in]       adc     ADC to get the highest supported resolution of
+ */
+static inline uint8_t adc_ng_min_res(uint8_t adc)
+{
+    assert(adc < ADC_NG_NUMOF);
+    return (uint8_t)bitarithm_lsb(adc_ng_drivers[adc]->res_supported);
+}
+
+/** @} */
+
+/**
+ * @brief   Initialize and power up the ADC channel @p chan of device @p adc
+ *
+ * @param[in]       adc     ADC device to initialize a channel of
+ * @param[in]       chan    Channel to initialize
+ * @param[in]       res     Resolution to sample at
+ * @param[in,out]   ref     Reference voltage in mV to use (***see details!***)
+ *
+ * @retval  0               Success
+ * @retval  -ENOTSUP        Requested resolution not supported
+ * @retval  -ENXIO          No such channel
+ * @retval  -ERANGE         Requested reference voltage is higher than all
+ *                          available references
+ * @retval  -EALREADY       The ADC is already powered and configured
+ * @retval  <0              A driver specific error occurred
+ *
+ * @note    On success, the ADC is powered up and consumes additional power
+ *          until @ref adc_ng_off is called.
+ *
+ * The reference voltage to use is given with @p ref in millivolt. The driver
+ * will pick a reference voltage that is as close to @p ref as possible, but
+ * not smaller. The actually chosen reference voltage is stored in @p ref.
+ */
+int adc_ng_init(uint8_t adc, uint8_t chan, uint8_t res, uint16_t *ref);
+
+/**
+ * @brief   Turn of the given ADC device
+ *
+ * @param[in]       adc     ADC device to turn off
+ */
+static inline void adc_ng_off(uint8_t adc)
+{
+    assert(adc < ADC_NG_NUMOF);
+    adc_ng_drivers[adc]->off(adc_ng_handles[adc]);
```

Comment:
Do we assume that the driver always provides an `off`-Funktion even if it is not supported?

Reply:
No, we should adapt this to check if the pointer is `NULL` :-)

Reply:
Done :-)